src/ControllerStatusIO.c

```diff
@@ -751,3 +763,86 @@ int Ros_Controller_GetActiveAlarmCodes(USHORT active_alarms[MAX_ALARM_COUNT + MA
 
     return num_entries;
 }
+
+/**
+ * Attempt to load on-controller extrinsic kinematic calibration data to improve
+ * accuracy of TF broadcasts (for frames for which such calibration data exists).
+ *
+ * NOTE: returns TRUE IFF at least one calibration file loaded successfully,
+ *       FALSE in all other cases.
+*/
+static BOOL Ros_Controller_LoadGroupCalibrationData(Controller* const controller)
+{
+    BOOL bOneCalibFileLoaded = FALSE;
+
+    for (int i = 0; i < MAX_ROBOT_CALIBRATION_FILES; i += 1)
+    {
+        MP_RB_CALIB_DATA calibData;
+        if (Ros_mpGetRobotCalibrationData(i, &calibData) == OK)
+        {
+            Ros_Debug_BroadcastMsg("%s: file %d loaded OK", __func__, i);
+
+            //Take note at least one calibration file loaded successfully.
+            //
+            //NOTE: successful loading of a single calibration file will of
+            //course not mean all groups (that would benefit from it) are actually
+            //calibrated correctly. But this is the best we can do, as there is
+            //no requirement to calibrate every group to every other group, nor
+            //will this prevent MotoROS2 from successfully controlling motion.
+            //
+            //TF frame broadcasts for uncalibrated groups will however be incorrect,
+            //which is why posting an alarm in case there are no calibration files
+            //at all on a multigroup system is considered a good thing to do.
+            bOneCalibFileLoaded = TRUE;
+
+            if (calibData.s_rb.grp_no <= MP_R8_GID && //the slave is a robot
+                calibData.m_rb.grp_no <= MP_R8_GID) //the master is another robot's RF
+            {
+                int groupIndex = mpCtrlGrpId2GrpNo((MP_GRP_ID_TYPE)calibData.s_rb.grp_no);
+                MP_COORD* coord = &controller->ctrlGroups[groupIndex]->robotCalibrationToBaseFrame;
+                coord->x = calibData.pos_uow[0];
+                coord->y = calibData.pos_uow[1];
+                coord->z = calibData.pos_uow[2];
+                coord->rx = calibData.ang_uow[0];
+                coord->ry = calibData.ang_uow[1];
+                coord->rz = calibData.ang_uow[2];
+            }
+        }
+    }
+
+    return bOneCalibFileLoaded;
+}
+
+//TODO(gavanderhoorn): should be able to make this 'static' once Yaskawa-Global/motoros2#173
+//is implemented and we can unit test static functions
+BOOL Ros_Controller_ShouldWarnNoCalibDataLoaded(Controller const* controller, BOOL bCalibLoadedOk, BOOL bPublishTfEnabled)
+{
+    size_t numRobotsOrStations = 0;
+    for(size_t i=0; i < controller->numGroup; ++i)
+        if (Ros_CtrlGroup_IsRobot(controller->ctrlGroups[i]) || Ros_CtrlGroup_IsStation(controller->ctrlGroups[i]))
+            numRobotsOrStations += 1;
```

Comment:
I don't remember exactly any more why I'm counting robot groups here as well again -- we already have `Controller::numRobot` ..

Reply:
Robots and stations have an explicit calibration file defined. The base axis calibration is done via robot parameters.

For this alarm to occur, we need to have multiple groups. But the base axes don't count. 

I think you should keep `numRobotsOrStations` and remove `numRobot`.